internal/configs/state_store.go

```diff
@@ -178,34 +183,35 @@ func (b *StateStore) Hash(stateStoreSchema *configschema.Block, providerSchema *
 			diags = diags.Append(diag)
 		}
 		if diags.HasErrors() {
-			return 0, 0, diags
+			return 0, diags
 		}
 	}
 
-	// We're on the happy path, so continue to get the hash
+	// We're on the happy path, but handle if we got a nil value above
 	if ssVal == cty.NilVal {
 		ssVal = cty.UnknownVal(schema.ImpliedType())
 	}
-	ssToHash := cty.TupleVal([]cty.Value{
-		cty.StringVal(b.Type),
-		ssVal,
-	})
 
 	// 2. Prepare the provider hash
 	schema = providerSchema.NoneRequired()
 	spec = schema.DecoderSpec()
 	pVal, decodeDiags := hcldec.Decode(b.Provider.Config, spec, nil)
 	if decodeDiags.HasErrors() {
 		diags = diags.Append(decodeDiags)
-		return 0, 0, diags
+		return 0, diags
 	}
 	if pVal == cty.NilVal {
 		pVal = cty.UnknownVal(schema.ImpliedType())
 	}
-	pToHash := cty.TupleVal([]cty.Value{
-		cty.StringVal(b.Type),
-		pVal,
-	})
 
-	return ssToHash.Hash(), pToHash.Hash(), diags
+	toHash := cty.TupleVal([]cty.Value{
+		cty.StringVal(b.Type), // state store type
+		ssVal,                 // state store config
+
+		cty.StringVal(b.ProviderAddr.String()),            // provider source
+		cty.StringVal(stateStoreProviderVersion.String()), // provider version
+		cty.StringVal(b.Provider.Name),                    // provider name - this is directly parsed from the config, whereas provider source is added separately later after config is parsed.
+		pVal,                                              // provider config
+	})
```

Comment:
Behold the beautiful new hash contents ✨

Reply:
That's great in terms of comments and clarity 👍🏻 🌟 

I am wondering why this is a tuple and why it contains just the values and not the attribute names but I think that is somewhat irrelevant to this PR. Just thinking out loud as I read through the code. 😅

Reply:
Yeah this is something that I've copied from the original implementation of hashing for `backend`s and haven't dug into that deeply. I can imagine the reason being that attribute names would be the same each time and not help with producing a unique hash per backend/state_store, and attribute names were less useful for hashing the 2 values from `backend`s, too.